src/compound/tagged_union.ts

```diff
@@ -0,0 +1,77 @@
+import { u8 } from "../primitives/mod.ts";
+import {
+  AlignedType,
+  type InnerType,
+  type Options,
+  type Packed,
+  type ValueOf,
+} from "../types/mod.ts";
+import { getBiggestAlignment } from "../util.ts";
+
+type Fn<V> = (variant: V) => InnerType<AlignedType<number>>;
+
+export class TaggedUnion<
+  T extends Record<number, AlignedType<unknown>>,
+  V extends ValueOf<{ [K in keyof T]: InnerType<T[K]> }> = ValueOf<
+    { [K in keyof T]: InnerType<T[K]> }
+  >,
+> extends AlignedType<V> implements Packed<V> {
+  #record: T;
+  #variantFinder: Fn<V>;
+  #discriminant: AlignedType<number>;
+
+  constructor(
+    input: T,
+    variantFinder: Fn<V>,
+    discriminantCodec: AlignedType<number> = u8,
```

Comment:
```suggestion
    discriminantCodec: AlignedType<number | string | symbol> = u8,
```
That way anything that can be used as a key in a record can be a discriminant?

Reply:
Sounds like a good idea. But I don't think we should add `symbol` because symbols are usually used in internal parts of the runtime or framework
Edit: Currently only custom components can return a symbol, nothing in byte_type implements it